apps/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts

```diff
@@ -0,0 +1,588 @@
+          }
+        }
+      );
+
+      // Ensure that node_modules folders exist after install, since the timestamps on these folders are used
+      // to determine if the install can be skipped
+      const projectNodeModulesFolders: string[] = [
+        path.join(this.rushConfiguration.commonTempFolder, RushConstants.nodeModulesFolderName),
+        ...this.rushConfiguration.projects.map((x) => {
+          return path.join(x.projectFolder, RushConstants.nodeModulesFolderName);
+        })
+      ];
+
+      for (const nodeModulesFolder of projectNodeModulesFolders) {
+        FileSystem.ensureFolder(nodeModulesFolder);
+      }
+    } catch (error) {
+      // All the install attempts failed.
+
+      if (
+        this.rushConfiguration.packageManager === 'pnpm' &&
+        this.rushConfiguration.pnpmOptions.pnpmStore === 'local'
+      ) {
+        // If the installation has failed even after the retries, then pnpm store may
+        // have got into a corrupted, irrecoverable state. Delete the store so that a
+        // future install can create the store afresh.
+        console.log(colors.yellow(`Deleting the "pnpm-store" folder`));
+        this.installRecycler.moveFolder(this.rushConfiguration.pnpmOptions.pnpmStorePath);
+      }
+
+      throw error;
+    }
+
+    console.log('');
+
+    // We need to update the repo state with the information from the install
+    const repoState: RepoStateFile = this.rushConfiguration.getRepoState(this.options.variant);
+    const commonVersions: CommonVersionsConfiguration = this.rushConfiguration.getCommonVersions(
+      this.options.variant
+    );
+    repoState.preferredVersionsHash = commonVersions.preferredVersionsHash;
+    if (repoState.saveIfModified()) {
+      console.log(
+        colors.yellow(
+          `${RushConstants.repoStateFilename} has been modified and must be committed to source control.`
+        )
+      );
+      console.log('');
+    }
+  }
+
+  /**
+   * Preferred versions are supported using pnpmfile by substituting any dependency version specifier
+   * for the preferred version during package resolution. This is only done if the preferred version range
+   * is a subset of the dependency version range. Allowed alternate versions are not modified. The pnpmfile
+   * shim will subsequently call into the provided pnpmfile, if one exists.
+   */
+  protected async createShimPnpmfile(filename: string): Promise<void> {
+    // Get the versions we want to add to the shim pnpmfile
+    const allPreferredVersions: Map<string, string> = InstallHelpers.collectPreferredVersions(
+      this.rushConfiguration,
+      this.options
+    );
+    const allowedAlternativeVersions: Map<
+      string,
+      ReadonlyArray<string>
+    > = this.rushConfiguration.getCommonVersions(this.options.variant).allowedAlternativeVersions;
+
+    const clientPnpmfileName: string = 'clientPnpmfile.js';
+    const pnpmfileContent: string[] = [
```

Comment:
Would it make more sense to write this file in TypeScript and just copy the built JavaScript to the `common/temp` folder along with a JSON file containing the data it needs?

Reply:
I had contemplated doing that. My first approach was to actually do a `function() { ... }.ToString()` for it, which is a bit of a hack. But one thing I couldn't think of a way to work around was locating the `semver` package that Rush uses and inserting that into the require statement.

Any thoughts?

Reply:
My primary concern is that this is a fair amount of code managed as just strings. It'd be good for this to be written in TypeScript and compiled. Maybe do that and have a placeholder where the data gets injected?

Reply:
+1 It might be small today, but people are likely to keep adding stuff to it over time.